### ISSUE TITLE
Training score dropped

### ISSUE THREAD

**Issue by PonteIneptique** [mittagessen/kraken#421]
As a follow-up to #420, I wanted to check what was going on.

Recap:

Between two different version of kraken, but mostly with different version of the deps, I lost 6-10 points of % on dev.

I saw you committed a potential fix here https://github.com/mittagessen/kraken/commit/33c1875d6f438cbb38ec10ff100586bcc0cd1d81

I just launched a check experiment to see if we reach the same dev score with the fix, and thought a separate issue would be better (as ReduceLRonplateau still crashes)

**Comment by colibrisson** [comment]
Could you share your training arguments?

**Comment by mittagessen** [comment]
Yes, please see if that optimizer return format change fixes things.

**Comment by colibrisson** [comment]
>
> please see if that optimizer return format change fixes things.
>

Lightning documentation says that ReduceLROnPlateau requires a monitor:

```
# The ReduceLROnPlateau scheduler requires a monitordef
configure_optimizers(self):
    optimizer = Adam(...)
    return {
        "optimizer": optimizer,
        "lr_scheduler": {
            "scheduler": ReduceLROnPlateau(optimizer, ...),
            "monitor": "metric_to_track",
            "frequency": "indicates how often the metric is updated"
            # If "monitor" references validation metrics, then
"frequency" should be set to a
            # multiple of "trainer.check_val_every_n_epoch".
        },
    }
```

**Comment by PonteIneptique** [comment]
I opened the issue to keep track of the specific scoring issue.

Here is the command 

`env/bin/ketos train -u NFD --device cuda:0 --augment -f binary -t ./manifest-A-train.txt -e manifest-A-dev.txt -B 16 --fixed-splits --lrate 1e-3 -o model-TEST --spec "[1,120,0,1 Cr4,2,32,4,2 Gn32 Cr4,2,64,1,1 Gn32 Mp4,2,4,2 Cr3,3,128,1,1 Gn32 Mp1,2,1,2 S1(1x0)1,3 Lbx256 Do0.5 Lbx256 Do0.5 Lbx256 Do0.5]"  --lag 15`

This command does drop significantly (7 points, 82ish vs 89ish on dev) on master from today (so your fix @mittagessen does not change things).

**Comment by dstoekl** [comment]
I usually use --lrate 0.0001 if -B =1 for this type of mode. If you have a batchrate higher than 1 did you try to increase the lrate by sqroot(B)?, i.e. 0.0004 in your instance?

**Comment by PonteIneptique** [comment]
The same command, except for reduceonplateau, on a previous environment
does reacher 7 points higher accuracy. I doubt it's related to anything
else than either reduceonplateau (which Ben doubts is the case) or a
dependency gone rogue.

**Comment by dstoekl** [comment]
pls try adapting lrate to 0.0001 or 0.0004. It helped in my cases.

**Comment by PonteIneptique** [comment]
@dstoekl The same command with the same data yields different results multiple time, I would bet this has nothing to do with LR (and if I remember well, I actually tested lower LR in the first tries before I "debugged" it to be a deps/version issue) 

@mittagessen Training is underway with lightning 1.7.7 and latest master. So I downgraded below 1.8, we'll see if that fixes it.

**Comment by colibrisson** [comment]
>
> Training is underway with lightning 1.7.7 and latest master.
>

Does `ReduceLROnPlateau` work with PL 1.7.7 ?

**Comment by PonteIneptique** [comment]
> Does `ReduceLROnPlateau` work with PL 1.7.7 ?

I wanted to try first without it, to see if the issue was indeed PL or the fact I was not using --schedule reducelronplateau.

**Comment by PonteIneptique** [comment]
I can confirm that ROL does not work on 1.7.7 with latest master. But latest master does not work with 1.7.5 either, which worked with Kraken 4.1.2

**Comment by PonteIneptique** [comment]
FYI, I went back to using 4.1.2 with Pytorch-lightning 1.8.3.
ReduceOnPlateau works over there while none of 1.7.7 and 1.8.3 worked with Kraken on master / 4.2.0

There might have been a regression somewhere, which might impact more than the LRScheduler.

**Comment by PonteIneptique** [comment]
Ok, I can confidently say that it is neither the `--schedule reduceonplateau` or `pytorch-lightning`, but something in Kraken in between version or another dependency.

On 4.1.2, with PL 1.8,3, I reach .86 at epoch 6 (so ROP could not be activated).

**Comment by PonteIneptique** [comment]
The only different thing I can see that could affect training / recognition is 

- The way metrics are logged ( https://github.com/mittagessen/kraken/compare/4.1.2...master#diff-974b9dde6ef2001d713a0eaaeb9bb2bd5f0f9c67e10eb1fb6d3aa1ccebee9701R165-R167 )
- The way optimizer and scheduler are set-up or stepped: https://github.com/mittagessen/kraken/compare/4.1.2...master#diff-974b9dde6ef2001d713a0eaaeb9bb2bd5f0f9c67e10eb1fb6d3aa1ccebee9701R576-R602

I cannot see anything else in the code that would relate to this. I feel like maybe what's hurting the ROP might also hurt training, and fixing one will fix the other ? (finger crossed ?)

**Comment by PonteIneptique** [comment]
So, to recap

| Kraken | PL | ReduceOP | Score |  
| ------ | ------ | ----- | ----- | 
| Master | 1.8.4 | Crash | 82 |
| Master | 1.7.7 | Crash | 82 |
| 4.1.2 | 1.8.4 | Works | 89 |
| 4.1.2 | 1.7.7 | Works | 89 |

**Comment by PonteIneptique** [comment]
So ROP does not seem to be tied to the issue: I "fixed" it and training seems to still behave weirdly (any training on 4.1.2 reachs 80+ acc at epoch 5, it went up and dived down at epoch 6 on master with ROP fix, cf. https://github.com/mittagessen/kraken/issues/420#issuecomment-1407754951 )

**Comment by colibrisson** [comment]
I guess the drop at epoch 6 is due to the scheduler kicking in. By default,
`rop_patience` is set to 5 so, if you only monitor the loss after each
epoch (which IMO doesn't make sense for large datasets), the scheduler has
to wait for 5 epochs before actually reducing the LR.

**Comment by PonteIneptique** [comment]
> I  guess the drop at epoch 6 is due to the scheduler kicking in.

But that would not make sense, would it ? or it would indicate that ROP in itself is broken ? Because acc was raising for the first five epochs, but dropped by 2 points (from what I remember epoch 5 was 79% and epoch 6 was 77%).

**Comment by colibrisson** [comment]
Maybe you got stuck in a saddle point after the scheduler kicked in. I find
it difficult to troubleshoot this issue without looking at the loss and LR
graphs.

**Comment by PonteIneptique** [comment]
AFAIK, Schedulers monitor `accuracy` and not loss in Kraken, which would mean it had no reason to kick in as it was not on a downward slope, no ?

**Comment by PonteIneptique** [comment]
This is accuracy plotting, I am running another test right now for longer on the fixed ROP.

![image](https://user-images.githubusercontent.com/1929830/215424907-d91e5df4-7b2a-42a8-a8b0-2e91a863e690.png)

Top models are 4.1.2 with different setting

**Comment by PonteIneptique** [comment]
(I have started monitoring the loss to compare between the two.)

**Comment by PonteIneptique** [comment]
I'll provide both arrows soon. I think I cannot really go further right now, mostly including the electric bill I am gonna get by testing this.

One thing which I have not checked is if the data are somehow treated differently, despite using the same arrow (empty lines ? encoding ?)

**Comment by PonteIneptique** [comment]
@mittagessen I am available for a call if you wanna talk about this ;)

**Comment by mittagessen** [comment]
Thanks for the work. I've got the French state paying for the electricity bill so I'll run some tests later. Can you send me the dataset? It would make it easier to build a test case without me having to find some training data as my collection got nuked.

The ROP only works on the validation metric (accuracy in this case), the loss is completely ignored. You could plot it but it usually isn't terribly indicative of anything (that's why I haven't spend the time to get it to be printed out on the progress bar because PTL makes that weirdly difficult).

**Comment by PonteIneptique** [comment]
It's currently being uploaded :) I am on a research trip in Poitiers, which makes connection quite... slow :)

**Comment by mittagessen** [comment]
Okidokey. I'll fix the ROP issue in the meanwhile.

**Comment by PonteIneptique** [comment]
Have fun ! https://drive.google.com/file/d/1w-OUa3W9uNCAYbsHhHoDff03vWXSoP8i/view?usp=sharing

**Comment by mittagessen** [comment]
Thanks. I've got a theory on why stuff might be broken. We're using manual optimization (for the learning rate scheduling/warmup necessary for pretraining) from 4.2.0 but don't actually set the `automatic_optimization` flag in the trainer's `__init__`. It's possible this causes aberrant behavior like the optimizer being called twice or something.

**Comment by PonteIneptique** [comment]
Keep us updated :)

**Comment by mittagessen** [comment]
OK, I can reproduce the drop somewhere between 4.1.2 and current master (even with `automatic_optimization` set to `false`). That's at least something.

**Comment by PonteIneptique** [comment]
Thank you for confirming this, I was starting to doubt myself :)

**Comment by mittagessen** [comment]
Unfortunately it means bisecting through the tree backwards to find the regression. And I suspect it is in the large refactoring commit for the pretraining.

**Comment by PonteIneptique** [comment]
Yup... THe good thing is from what I saw, it is detectable with only few epochs (4 or 5)

**Comment by mittagessen** [comment]
Found the bad commit (ecb47081d64eb42fdb66ce344f26576ed54ab480). Unfortunately, it is the large pretraining merge one.

**Comment by mittagessen** [comment]
Found the source of the error. Empty lines in binary datasets don't get properly filtered in master but they are in 4.1.2. I'll push a fix tomorrow.

**Comment by PonteIneptique** [comment]
WOoohoo ! (For two reasons: one you found it, and two I started feeling like it might be based on data... :) )

Can I ask if you could publish a release after that ? :D

**Comment by mittagessen** [comment]
On 23/02/01 12:23PM, Thibault Clérice wrote:
> WOoohoo ! (For two reasons: one you found it, and two I started feeling like it might be based on data... :) )
> 
> Can I ask if you could publish a release after that ? :D 

Yes, it's one of the last blocking regressions for a new release. I'm
writing *slow* training tests to catch stuff like this and then tag one.

**Comment by mittagessen** [comment]
BTW you triggered this bug because your binary dataset contains a lot of lines that are transcribed as a single whitespace. These do get included in the compilation process (because they are not completely empty) but get squashed to empty strings by the default text processors in the dataset (and should therefore be filtered out). Adding 180 lines without good labels just breaks the model training.
